### ISSUE TITLE
Should variant `start` or `gene` be mandatory?

### ISSUE THREAD

**Issue by buske** [MatchMakerExchange/mme-apis#62]
From François:
`Not sure about this but does a variant without a start make sense ?`

I would tend to agree, but it depends on what we're planning on putting in the `variant` field versus the `gene` field. Originally, the idea was that the gene symbol might be as specific as we wanted to localize the mutation. Do we still want to be able to specify that a gene has (e.g.) a het de-novo stopgain mutation, without saying what the actual mutation is? If so, I think we either need to add `gene symbol` to the `variant` field (somewhat counterintuitive), or add `zygosity` and `type` to the `gene` field (somewhat redundant) and make `start` be mandatory.

**Comment by Relequestual** [comment]
I agree varinats should have a manditory start and end.

After reviewing these changes, it reminded me of a situation which would suggest we should add gene symbol to the variant.

When users input variants, they are required to also include an ensembl transcript. We use this to determin the gene name. There are instances where without the transcript data, it is not certian which gene the variant effects. It could be useful to someone viewing the variant to know which gene the variant is effecting.

I would suggest, as an additional change, to include the gene name for the variant (maybe this could be optional). We would then be able to return only patients who variants in the defined location also match the gene.

**Comment by buske** [comment]
Regarding `end` being mandatory: My understanding is that `end` is only useful for larger variants (e.g. CNVs) where you might not want to specify the whole sequence in `referenceBases`. Otherwise, `end` can be directly computed from the `start` and `referenceBases.length`. We could make both required if this doesn't bother us.

Regarding adding a `gene` to each variant object: could you clarify the use case a bit more (since the gene is basically redundant with variant position)? If we add gene info back into variants, would it be useful to reconsider merging the candidate genes and variants back together again, so that we're just representing the candidate genetics relevant for matching? I was hoping to separate out "raw" variants from the candidate genes, but maybe this isn't best.

**Comment by cmungall** [comment]
@buske "Do we still want to be able to specify that a gene has (e.g.) a het de-novo stopgain mutation, without saying what the actual mutation is"

I believe the G2P WG had the same requirement (not possible with the current avro schema), @nlwashington may know more

**Comment by fschiettecatte** [comment]
GeneMatcher data contains gene identifier, chromosome, start, end and build. Gene identifier is required, end is optional, start and build are required if chromosome is set. The new schema requires me to split out gene identifier and variants info, so no issue for me if we re-add in the gene identifier with the variant. Requiring an end is not an issue,I would just sent it to the start if it is not in the data.

**Comment by Relequestual** [comment]
I mistakenly suggested `end` should be manditory... apologies.

Regarding `gene` being in the variant level data:
We have seen situations where an SNV location wise covers multiple genes. Somtimes the genes are on opposite strands, or simply overlapping. We store a transcript with each variant, which allows us to determin the gene the variant effects.
The reciving system can't check form themselves which gene by the same method, because they won't have the selected transcript.
I can't remember the specific example, but by specifying the gene if known, could result in better matches, as if gene A is specified, and a potential result specifies gene B, it would not be included in the returning results.
I know it's unliekly, but it's something I've seen and had to have explained to me before.

**Comment by buske** [comment]
I think I understand what you mean. Would it be fair to summarize this as: the `type` argument can be ambiguous without a corresponding `gene` field?

**Comment by Relequestual** [comment]
After discussion, I am happy that `gene` should not be mandatory. Forgot about CNVs.

`start` though should be mandatory.

My thinking was, we could provide better matching if the `gene` field is included, but it can't always be included.

I'll make a PR which changes `start` to mandatory.

**Comment by Relequestual** [comment]
This seems like it should be part of v1.0, as it's a result of the restructuring of objects.

**Comment by Relequestual** [comment]
I know I'm re-opening this issue, but I've come across a potetial problem moving forward.

Assuming we want to make matches as valuable as they can be, having the gene as a manditory field in the varinat, is again, useful. I'll explain the usecase.

We are discussing a better way to do internal matching. From a clincians perspective, I am told, they are more interested in finding a variant which overlaps the **same gene AND has the same functional consequence**. We currently don't require the consequence (we define it as `type`), and that's fine, but when it is provided, we can do a better match.

We can determin a set of genes which overlap the variant, however having the gene for the variant would allow better matching in the instance where an SNV overlaps several genes.

I don't know if it's viable to add a gene field for variants back, but with the genes being seperated from the varinats, there's no way to know which candiate gene is associated with which varinat if more than one variant is associated with the patient (we have patients who have a number of variants entered). I guess it depends on if there's a reasonable expectation that a candidate gene will be stored against a vairnat in most systems (I don't know).

Again just to calirfiy, these are variants manually chosen by clinicnas, not just any variant. In the instance where a patient has multiple vairants in our database, they are all considered of interest.

**Comment by fschiettecatte** [comment]
I have no problem with this, in fact GeneMatcher stores gene identifiers with the variants, and the separation in the spec always bothered me. I would prefer to have a single variants section. We could just make all the fields optional which would allow us to support all current functionality (we could require 'start' if 'referenceName' is specified, and required gene identifier if referenceName/start is specified.)

**Comment by buske** [comment]
@Relequestual I completely agree. It's related to the use case of specifying "a het stopgain variant in this gene"

@fschiettecatte By "single variants section" do you mean combining the candidate genes and variants back together into the genomicFeatures section with most of the things being optional?

Hmm, based on these comments, it sounds similar to the way we were approaching it before Miami. If that's the case, we could have the following instead of the separate `genes` and `variants` fields (where starred fields are required if the parent is provided).
```json
"genomicFeatures": [{
    "gene"* : {
        "id"*
    },
    "variant": {
        "assembly"*
        "referenceName"*
        "start"*
        "end"
        "referenceBases"
        "alternateBases"
    },
    "zygosity"
    "type": {
        "id"*
        "label"
    }
}]
```

The corresponding use cases would then be:
* Het stopgain variant in gene:
```json
{
  "gene" : {"id":"SRCAP"},
  "zygosity":1,
  "type":{"id":"SO:whatever", "label":"STOPGAIN"}
}
```
* Specific stopgain variant in gene:
```json
{
  "gene" : {"id":"SRCAP"},
  "variant" : {"assembly":"GRCh37", "referenceName":"chr16", "start":123456, "referenceBases":"C", "alternateBases":"T"}
  "zygosity":1,
  "type":{"id":"SO:whatever", "label":"STOPGAIN"}
}
```
* Candidate gene
```json
{
  "gene" : {"id":"SRCAP"}
}

**Comment by fschiettecatte** [comment]
Works for me, I assume the reason you want:
```
"gene"* : {
        "id"*
    },
```
is to be able to add additional attributes in the future.

**Comment by buske** [comment]
@fschiettecatte Yep, that was the idea, though I don't yet know what those would be. >.<

**Comment by fschiettecatte** [comment]
Great, I like your use of 'label' to add additional data, I have been prepending underscores to element names to do the same (see below). Is this something we want to document in the spec? I imagine this depends on whether the validator written by @Relequestual will barf on them. If not I would vote for the spec to be silent on that, i.e. you can add what you want for your own internal use but "you own it if you break it".
```
"genes": [
    {
        "id": "ENSG00000133392"
    }
], 
"variants": [
    {
        "_ensemblGeneID": "ENSG00000133392", 
        "_entrezGeneID": "4629", 
        "_geneName": "MYH11", 
        "assembly": "GRCh37", 
        "end": 15809106, 
        "referenceName": "16", 
        "start": 15809106
    }
]
```

**Comment by buske** [comment]
@fschiettecatte I like the idea of having the validation being silent on any extra fields, which matchmakers should be free to add as they see fit, but which they have to maintain and make sure they don't collide with the API in any unfortunate ways. :)

Regarding the underscores, I don't have an opinion either way, but I like the pythonic semantics as long as it's proper JSON. Could someone check? 

Based on your example, I do think that we should probably add a field to the `gene` object that says what source the gene ID corresponds to (e.g. 'HGNC', 'entrez', 'ensembl'), but I'm not sure the best way to support handling multiple `id`s. Others should probably be specified as internal or alternate IDs, so that there's always one primary? @cmungall any suggestions based on your experience with the same concept having many different mappings (but needing to specify at least one)? Perhaps something like:
```json
"genomicFeatures": [
    {"gene": {
        "id":"MYH11",
        "source":"HGNC",
        "alternateIds":[
            {"id":"ENSG00000133392", "source":"ensembl"},
            {"id":"4629", "source":"entrez"}
        ]
]
```

**Comment by fschiettecatte** [comment]
We need to simplify, because we need to ship, how about this:
```
"genomicFeatures": [
    {"gene": {
        "id":"HGNC:MYH11"
        }
    },...
 ```
or
```
"genomicFeatures": [
    {"gene": {
        "id":"ensembl:ENSG00000133392"
        }
    },...
 ```
or:
```
"genomicFeatures": [
    {"gene": {
        "id":"entrez:4629"
        }
    },...
```
This mirrors what was done for other IDs.

**Comment by buske** [comment]
Okay, good call, François. I got carried away (as usual). 

I think your solution is good, but I don't know what the proper curie prefixes for these would be, and it might take a bit to figure that out. "HGNC" for instance seems to collide with their numerical ID system, based on what I get when I look at (their export)[http://www.genenames.org/cgi-bin/download]. Unless someone can step in with a solution that has already worked for describing genes in this way in other systems, I'd suggest tabling the gene ID discussion for now, leaving it the way it currently is, and allowing you to add whatever optional fields you want in the meantime.

I'll do a `genes`/`variants` merge pull request later today. Hopefully that's it for v1.0 and then we can ship?

**Comment by fschiettecatte** [comment]
Yes, sounds good. Agreed that using 'HGNC:' is not good, this is compounded with the issue that gene symbols are not true IDs (they change and they move) but everyone uses them so we will need to support them, I think.

**Comment by buske** [comment]
Hmm, given that each system must implement their own gene ID mapping internally anyway (to handle matching given symbol, Entrez, or Ensembl, and to actually validate any user-entered genes), we could theoretically remove gene symbols from the API altogether. We could even standardize on one, e.g. Ensembl.

**Comment by fschiettecatte** [comment]
I would leave all three for now, different systems use different IDs, my experience with setting up the external links in OMIM is that Entrez Gene IDs and Gene Symbols are what people use. Ensembl Gene IDs are not as commonly used. In addition there isn't a 1-to-1 mapping between the IDs which further complicates things.

**Comment by buske** [comment]
@fschiettecatte Fair point. Agreed.

**Comment by Relequestual** [comment]
This looks good to me!
@fschiettecatte The json-schema validation can support additional fields, I just have to explicitly allow it. I can do so in this case! =]

Regarding Ensembl Gene IDs. It sure is strange that ensembl consideres two genes a synonym, while HGNC does not. How does this happen? Possibly different purposes? For example, what makes MIR548AA2 and MIR548D2 different?

(I've added +1 to the PR. One more required!)

**Comment by fschiettecatte** [comment]
@Relequestual I would not make any changes to the json-schema validation to support anything other than the official spec. If a site (like me) decides to add it extra fields then it is that site's responsibility to make sure it does not break other people's implementation.

As for MIR548AA2 vs. MIR548D2 you will need to put the question to HGNC and Ensembl.

**Comment by Relequestual** [comment]
@fschiettecatte Hum OK.

I actually did file a report with Ensembl directly (direct access to their helpdesk queue!). They said it's probably an issue with the pipeline for specific circemstances. Issue noted.

I don't think the intension is to not have a 1-to-1 mapping, so I'd report any issues like that you see.

**Comment by fschiettecatte** [comment]
@Relequestual a year ago I reported an issue where 1,700 of the Ensembl genes disappeared from the NCBI gene2ensembl file, still waiting on a fix :)

**Comment by Relequestual** [comment]
@fschiettecatte haha =[ Always feel free to pass Ensembl queries through me in future if you'd like =]

**Comment by Relequestual** [comment]
I'm calling this issue closed. If we need to / want to review gene notations, we can discuss on google groups or create a new issue =]
